### PR TITLE
cmd/geth, eth: added canonical extra data

### PR DESCRIPTION
**Pull request by obscuren**
Implemented canonical extra data according to
https://github.com/ethereum/wiki/wiki/Extra-Data

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Thu Aug 6 09:21:54 UTC 2015_

**Comment by bas-vk** [comment]
- Consider setting the encouraged extra data also in the console sub command
- The wiki page describes version 0 for the first version while your PR uses version 1. I'm not sure what the actual version should be.

Code looks ok.

**Comment by obscuren** [comment]
The number doesn't matter. I've changed this to reflect the version number of `geth`

**Comment by frozeman** [comment]
:+1:

File: cmd/geth/main.go [hunk 1]
```diff
@@ -346,6 +350,20 @@ func main() {
...
 }
 
+func makeDefaultExtra() []byte {
+	var clientInfo = struct {
+		Version   uint
+		Name      string
+		GoVersion string
+		Os        string
+	}{uint(VersionMajor<<16 | VersionMinor<<8 | VersionPatch), ClientIdentifier, runtime.Version(), runtime.GOOS}
+	extra, err := rlp.EncodeToBytes(clientInfo)
```
Comment (Gustav-Simonsson) [inline cmd/geth/main.go]: While this works now, we have no control over values returned from runtime, and while unlikely it's plausible they could return longer strings causing the RLP to be larger than 32 bytes, breaking mining for someone upgrading Go version.

Add a check to truncate strings to guarantee RLP is never more than 32 bytes.

Comment (obscuren) [inline cmd/geth/main.go]: Better yet validation needs to be added to `miner.SetExtra`. Adding right now.

**Comment by fjl** [comment]
:+1:

**Comment by Gustav-Simonsson** [comment]
:+1:

**Pull request merged**